examples/mauna-loa/script.jl

```diff
@@ -0,0 +1,236 @@
+ytest = co2[idx_test]
+
+function plotdata()
+    plot(; xlabel="year", ylabel="CO₂", legend=:bottomright)
+    scatter!(xtrain, ytrain; label="training data", ms=2, markerstrokewidth=0)
+    return scatter!(xtest, ytest; label="test data", ms=2, markerstrokewidth=0)
+end
+
+plotdata()
+
+# ## Prior
+#
+# We will model this dataset using a sum of several kernels which describe
+#
+# - smooth trend: squared exponential kernel with long lengthscale;
+# - seasonal component: periodic covariance function with period of one year,
+#   multiplied with a squared exponential kernel to allow decay away from exact
+#   periodicity;
+# - medium-term irregularities: rational quadratic kernel;
+# - noise terms: squared exponential kernel with short lengthscale
+#   and uncorrelated observation noise.
+#
+# For more details, see [Rasmussen & Williams (2005), chapter 5](http://www.gaussianprocess.org/gpml/chapters/RW5.pdf).
+
+# We will use
+# [ParameterHandling.jl](https://invenia.github.io/ParameterHandling.jl/) for
+# handling the (hyper)parameters of our model. It provides functions such as
+# `positive` with which we can put constraints on the hyperparameters, and
+# allows us to represent all required parameters as a nested NamedTuple:
+
+#! format: off
+## initial values to match http://stor-i.github.io/GaussianProcesses.jl/latest/mauna_loa/
+θ_init = (;
+    se1 = (;
+        σ = positive(exp(4.0)),
+        ℓ = positive(exp(4.0)),
+    ),
+    per = (;
+        σ = positive(exp(0.0)),
+        ℓ = positive(exp(1.0)),
+        p = fixed(1.0),  # 1 year, do not optimise over
+    ),
+    se2 = (;
+        σ = positive(exp(4.0)),
+        ℓ = positive(exp(0.0)),
+    ),
+    rq = (;
+        σ = positive(exp(0.0)),
+        ℓ = positive(exp(0.0)),
+        α = positive(exp(-1.0)),
+    ),
+    se3 = (;
+        σ = positive(exp(-2.0)),
+        ℓ = positive(exp(-2.0)),
+    ),
+    noise_scale = positive(exp(-2.0)),
+)
+#! format: on
+#md nothing #hide
+
+# We define a couple of helper functions to simplify the kernel construction:
+
+SE(θ) = θ.σ^2 * with_lengthscale(SqExponentialKernel(), θ.ℓ)
+## PeriodicKernel is broken, see https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/389
+##Per(θ) = θ.σ^2 * with_lengthscale(PeriodicKernel(; r=[θ.ℓ/2]), θ.p)  # NOTE- discrepancy with GaussianProcesses.jl
+function Per(θ)
+    return θ.σ^2 * with_lengthscale(SqExponentialKernel(), θ.ℓ) ∘ PeriodicTransform(1 / θ.p)
+end
+RQ(θ) = θ.σ^2 * with_lengthscale(RationalQuadraticKernel(; α=θ.α), θ.ℓ)
+#md nothing #hide
+
+# This allows us to write a function that, given the nested tuple of parameter values, constructs the GP prior:
+
+function build_gp_prior(θ)
+    k_smooth_trend = SE(θ.se1)
+    k_seasonality = Per(θ.per) * SE(θ.se2)
+    k_medium_term_irregularities = RQ(θ.rq)
+    k_noise_terms = SE(θ.se3) + θ.noise_scale^2 * WhiteKernel()
+    kernel = k_smooth_trend + k_seasonality + k_medium_term_irregularities + k_noise_terms
```

Comment:
```suggestion
    k_noise_terms = SE(θ.se3)
    kernel = k_smooth_trend + k_seasonality + k_medium_term_irregularities
```

The reason for suggesting this (and moving the noise to the FiniteGP) is that the semantics are subtley different, and the thing I'm proposing is what I _think_ we want. The `WhiteKernel` says "non zero covariance between f(x) and f(y) if x == y", while noise-in-FiniteGP says "cov(f(x), f(y)) = 0 always". The latter means that we can always safely add a diagonal matrix to a kernel matrix, while the former requires you to carefully check that no two inputs used to construct the kernel matrix are the same, and to add off-diagonal terms if they are the same.

I would rather that we encourage users to use the FiniteGP way of handling noise, because it's typically what we actually want.

(Also, the gradient w.r.t. inputs isn't defined for the kernel matrix associated with the `WhiteKernel`, but it works just fine with noise-in-FiniteGP)

Reply:
That's actually how I had it originally, but looking at Rasmussen&Williams again I just wanted to follow their example more closely... it somewhat changes the visualisation: using the WhiteKernel, its variance increases the uncertainty when plotting the posterior GP. Using the FiniteGP observation noise, we'd have to manually add the noise variance (`predict_y` in GPflow... was wondering if we want to support that in AbstractGPs directly, kinda goes more in the LatentGP direction though?)

Let me rephrase my understanding of the difference between WhiteKernel and noise-in-FiniteGP: in the former, covariance is finite only if xs[i] == xs[j] (based on the _values_ of the features), whereas in the latter, covariance is finite only for i == j (based on the _indices_). Does that match yours?

Reply:
> Let me rephrase my understanding of the difference between WhiteKernel and noise-in-FiniteGP: in the former, covariance is finite only if xs[i] == xs[j] (based on the values of the features), whereas in the latter, covariance is finite only for i == j (based on the indices). Does that match yours?

Yes. This matches mine. I also think of it as always getting "fresh randomness" with the noise-in-FiniteGP version.

> Using the FiniteGP observation noise, we'd have to manually add the noise variance

Yes, we would have to do this. I think it's worth doing to emphasise that our abstractions make this particular distinction.

> (predict_y in GPflow... was wondering if we want to support that in AbstractGPs directly, kinda goes more in the LatentGP direction though?)

I wouldn't be overly keen on having a `predict_y` function -- I would rather just ensure that `LatentGP`s with simple Gaussian likelihoods work nicely and are an option that people can easily use.

Reply:
So what would you tell someone (like this notebook) how to easily get the observation uncertainty included?

Maybe in some cases using the WhiteKernel like this _is_ the right thing to do? I'd be happy to add a paragraph to the notebook explaining the distinction...

> (Also, the gradient w.r.t. inputs isn't defined for the kernel matrix associated with the WhiteKernel, but it works just fine with noise-in-FiniteGP)

I just realised I'm not sure what you mean and/or why this is...

Reply:
> So what would you tell someone (like this notebook) how to easily get the observation uncertainty included?

I would tell them to pick out the marginals they're interested in in `fpost_opt` by writing `fpost_opt(x, noise)` and to plot that. It's obviously not as clean as the current thing, but I think it makes sense.

> Maybe in some cases using the WhiteKernel like this is the right thing to do? I'd be happy to add a paragraph to the notebook explaining the distinction...

That would be great.

> I just realised I'm not sure what you mean and/or why this is...

I mean that `d/dx k(x, y)` doesn't really exist when `x == y` if `k` is a `WhiteKernel`.